### PR TITLE
Add wildcard matching to dimension selection

### PR DESCRIPTION
**Pull request by fstehle**
Enhancement to PR https://github.com/prometheus/cloudwatch_exporter/pull/6
Add wildcard matching to dimension selection.

### CODE REVIEW

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 1]
```diff
@@ -180,7 +183,7 @@ private boolean useMetric(MetricRule rule, Metric metric) {
         String dimensionValue = dimension.getValue();
         if (dimensionSelectKeys.contains(dimensionName)) {
           List<String> allowedDimensionValues = rule.awsDimensionSelect.get(dimensionName);
-          if (!allowedDimensionValues.contains(dimensionValue)) {
```
Comment (brian-brazil) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java]: I think this needs a separate setting, as there's cases where we'd want to pass the aws_dimension_select directly to AWS (the billing API docs say it needs this for example).

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 2]
```diff
@@ -188,6 +191,38 @@ private boolean useMetric(MetricRule rule, Metric metric) {
...
+          return true;
+        }
+      }
+      return false;
+    }
+
+    /**
+     * Check if a string with wildcard characters matches a given input value
+     */
+    protected static boolean wildcardMatch(String pattern, String input) {
```
Comment (brian-brazil) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java]: Let's just go with regexes

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java (hunk 1 above)
Comment (fstehle) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java]: I do not understand the use case. 
Can you explain this in more detail please?

Comment (brian-brazil) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java]: http://docs.aws.amazon.com/AmazonCloudWatch/latest/DeveloperGuide/billing-metricscollected.html See the currency dimension.

**Comment by fstehle** [comment]
So I would suggest having an additional property `aws_dimension_select_regex`, eg.
```
"aws_dimension_select_regex": {
  "LoadBalancerName": ["prod-(.*)"]
}
```
If `aws_dimension_select_regex` and `aws_dimension_select` are set at the same time all dimensions which match the regex in `aws_dimension_select_regex` or the string in `aws_dimension_select` are returned.

**Comment by fstehle** [comment]
@brian-brazil Could you have a look on the updated commit please?

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 3]
```diff
@@ -168,12 +173,22 @@ private CloudWatchCollector(JSONObject config, AmazonCloudWatchClient client) th
     }
 
     /**
-     * Check if a metric should be used according to `aws_dimension_select`
+     * Check if a metric should be used according to `aws_dimension_select` or `aws_dimension_select_regex`
      */
     private boolean useMetric(MetricRule rule, Metric metric) {
-      if (rule.awsDimensionSelect == null) {
+      if (rule.awsDimensionSelect == null && rule.awsDimensionSelectRegex == null) {
```
Comment (brian-brazil) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java]: Will this cause an NPE is only one is null?

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 4]
```diff
@@ -168,12 +173,22 @@ private CloudWatchCollector(JSONObject config, AmazonCloudWatchClient client) th
...
+        return true;
+      }
+      if (metricsIsInAwsDimensionSelect(rule, metric) || metricIsInAwsDimensionSelectRegex(rule, metric)) {
```
Comment (brian-brazil) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java]: This should be an and I'd say (not that it'd make much sense - maybe forbid that as a config error?)

**Comment by brian-brazil** [comment]
Can you add unittests please?

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java (hunk 4 above)
Comment (fstehle) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java]: Why should we disallow that users can specify some AWS metrics with an exact string match and some metrics with an regex match? 

For example the following configuration to get all load balancers metrics with the name `myelb1`, `myelb2` or the prefix `anotherelb`:
```
{ 
  "region": "eu-west-1",
  "delay_seconds": 600,
  "metrics": [
    {
      "aws_namespace": "AWS/ELB", "aws_metric_name": "HealthyHostCount",
      "aws_dimensions": ["LoadBalancerName"], "aws_statistics": ["Minimum"],
      "aws_dimension_select": {
        "LoadBalancerName": ["myelb1", "myelb2"]
      },
      "aws_dimension_select_regex": {
        "LoadBalancerName": ["anotherelb(.*)"]
      },
      "period_seconds": 300
    }
  ]
}
```

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java (hunk 3 above)
Comment (fstehle) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java]: You are right. I will fix that.

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java (hunk 4 above)
Comment (brian-brazil) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java]: I could see that be confusing - is that an AND or and OR? Given how the rest works I'd expect it to be an AND.

Comment (fstehle) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java]: I changed this.
Now it is forbidden to  set `aws_dimension_select` and `aws_dimension_select_regex` at the same time:
https://github.com/Jimdo/cloudwatch_exporter/blob/dimension_value_wildcard/src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java#L110

**Comment by fstehle** [comment]
I added a unittest for it.

**Comment by fstehle** [comment]
@brian-brazil Could you please have another look at this PR?

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**